### ISSUE TITLE
Issue with loading models

### ISSUE THREAD

**Issue by hjlee9295** [amaiya/ktrain#290]
Hi, 

I saved and downloaded my predictor from google colab, and then loaded it with

`loaded_predictor = ktrain.load_predictor(fpath)`

When I run the model on my local machine for the web app (I am using streamlit for deployment), it works fine.
However, when I try to deploy through share.streamlit.io, I get this error:

Could not find a .preproc file in either the post v0.16.x loction (/app/mbti_twitter/models/tf_model.preproc) or pre v0.16.x location (/app/mbti_twitter/models.preproc)

At first, I thought the file path was wrong, so I double-checked that the file path is correct and the file exists.

Then, I tried to separately load models:

`import joblib`
`import ktrain`
`from ktrain import text`

`preproc = joblib.load(preproc_name)`
`model = preproc.get_model(fpath=fpath)`
`predictor = text.predictor.TextPredictor(model, preproc, batch_size=32)`

This piece of code also works fine on my local machine, yet it's not working on the web.
I thought it was an issue with streamlit, so I tried to deploy it with Heroku, but I am getting the same error message which is: 

![Screen Shot 2020-12-07 at 10 44 41 PM](https://user-images.githubusercontent.com/17819888/101437962-4ddb1d00-38df-11eb-98f6-8ec8890185ce.png)

I have been stuck on this issue for a while, and any help would be appreciated!

Thank you in advance.

**Comment by amaiya** [comment]
Hello,

I was able to deploy to Heroku with no problem using flask using the code example included in this [FAQ entry](https://github.com/amaiya/ktrain/blob/master/FAQ.md#how-do-i-deploy-a-model-using-flask).  If you're using different versions of Python or other libraries, try making them consistent.  I was using Python 3.6 both locally and on Heroku.  You can change the Python version on Heroku by including a `runtime.txt` file.

That said, when deploying models, you might also consider converting the model to ONNX (with optional quantization).  This can drastically reduce the footprint of your model to keep your model under the 500MB slug size limit on Heroku.  See [this  example notebook](https://github.com/amaiya/ktrain/blob/master/examples/text/ktrain-ONNX-TFLite-examples.ipynb) for converting **ktrain**-trained models to ONNX.  This allows you to make predictions without **ktrain** or even **tensorflow** installed.  If you're trying to deploy a Hugging Face **transformers** model like DistilBert, it may be beneficial to convert your model to PyTorch and then to ONNX.  The reason is that ONNX models converted from PyTorch can accept dynamically-sized input - no padding needed so faster predictions.  I recently did this when deploying a **ktrain**-trained model to Heroku and it worked quite well.  See [this FAQ entry](https://github.com/amaiya/ktrain/blob/master/FAQ.md#how-do-i-make-quantized-predictions-with-transformers-models) for how to do this with your own model.

**Comment by hjlee9295** [comment]
Thank you for the response!

I will work on the issue with suggested solutions for the next couple of days and follow up with the update!

**Comment by hjlee9295** [comment]
Hi,

I tried to convert the saved model to both TF Lite and ONNX, and both of them didn't work.

Both methods require me to load predictor first as below, and I cannot load my predictor when I deploy models on the server for the same reason I described above. 

`predictor = ktrain.load_predictor('/tmp/my_distilbert_predictor')`

Is there any other way to upload a model than using ktarin.load_predictor function?

Thank you for your help in advance!

**Comment by amaiya** [comment]
You don't need `load_predictor` or **ktrain** installed on the deployment machine when using ONNX.  As mentioned in [this FAQ](https://github.com/amaiya/ktrain/blob/master/FAQ.md#how-do-i-make-quantized-predictions-with-transformers-models), the `Predictor` references can easily be replaced.   In fact, I'm running two models right now with only ONNX installed (no ktrain, no TensorFlow, and no PyTorch).  I'll try to make this more clear in the notebook in the future.   In any case, once the model is converted to ONNX, you can just create a a session and make predictions:

```python
def create_onnx_session(onnx_model_path):
    provider='CPUExecutionProvider'
    from onnxruntime import GraphOptimizationLevel, InferenceSession, SessionOptions, get_all_providers
    assert provider in get_all_providers(), f"provider {provider} not found, {get_all_providers()}"
    options = SessionOptions()
    options.intra_op_num_threads = 0
    options.graph_optimization_level = GraphOptimizationLevel.ORT_ENABLE_ALL
    session = InferenceSession(onnx_model_path, options, providers=[provider])
    session.disable_fallback()
    return session

sess = create_onnx_session('model.onnx')

# class_names is list of label names
# maxlen is maximum length (e.g., 128)
tokens = tokenizer.encode_plus('My computer monitor is blurry.', max_length=maxlen, truncation=True)
tokens = {name: np.atleast_2d(value) for name, value in tokens.items()}
print(class_names[np.argmax(sess.run(None, tokens)[0])])

```
The `create_onnx_session` function is simply copied from `predictor.py` in **ktrain**.
You just need ONNX and `transformers` installed:
```python
!pip install --upgrade onnxruntime==1.5.1 onnxruntime-tools onnx transformers
```

I'll close this issue, but feel free to reply if you have further questions on getting this to work.

**Comment by hjlee9295** [comment]
Thank you for the quick reply! I will let you know how it goes! 👍

**Comment by hjlee9295** [comment]
Hi,

I am still struggling to deploy my app. I got onnx version working fine on my local machine, but again, when I tried to deploy it through heroku, I got the following error message. 

<img width="710" alt="Screen Shot 2020-12-13 at 12 43 23 PM" src="https://user-images.githubusercontent.com/17819888/102019463-be4fc700-3d41-11eb-9d4b-dd0b5b668fb0.png">

I've tried to search through online for troubleshooting but had not luck... Any advice would be very helpful!

Thanks!

**Comment by amaiya** [comment]
Hello:  Some things to try:
1. Make sure the version of libraries are exactly the same
2. Make sure the model file is transferred correctly (compare MD5s) - although not sure this would affect Heroku.

I've deployed ktrain/ONNX models to Heroku with no problem.  But, I used `fastapi`, so not sure if that will also make a difference.
